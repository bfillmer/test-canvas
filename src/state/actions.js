// @flow
import {createAction} from 'redux-actions'

// REDUX TYPES
export const types = {
  DRAG_START: 'svg/DRAG_START',
  DRAG_END: 'svg/DRAG_END',
  DRAGGED: 'svg/DRAGGED',
  ZOOMED: 'svg/ZOOMED'
}

// FLOW TYPES
type DragStartAction = {
  type: string,
  payload: {
    x: number,
    y: number,
    id: string
  }
}

type DragEndAction = {
  type: string
}

type DraggedAction = {
  type: string,
  payload: {
    x: number,
    y: number
  }
}

type ZoomedAction = {
  type: string,
  payload: {
    dy: number,
    id: string
  }
}

const startDrag: DragStartAction = createAction(types.DRAG_START, (x: number, y: number, id: string) => ({x, y, id}))
const endDrag: DragEndAction = createAction(types.DRAG_END)
const dragging: DraggedAction = createAction(types.DRAGGED, (x: number, y: number) => ({x, y}))
const zooming: ZoomedAction = createAction(types.ZOOMED, (dy: number, id: string) => ({dy, id}))

// ACTIONS
export const actions = {
  startDrag,
  endDrag,
  dragging,
  zooming
}
