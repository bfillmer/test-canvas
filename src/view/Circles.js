
import React from 'react'
import {connect} from 'react-redux'

const mapStateToProps = state => ({
  circles: state.svg.filter(s => s.id !== 'viewport')
})

const Container = ({circles}) => {
  const renderCircles = circles.map(s => <circle key={s.id} id={s.id} transform={`matrix(${s.matrix.join(' ')})`} r='20' fill='teal' stroke='black' />)
  return (<g>{renderCircles}</g>)
}

export const Circles = connect(mapStateToProps)(Container)
