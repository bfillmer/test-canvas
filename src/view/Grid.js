
import React from 'react'
import {connect} from 'react-redux'

const mapStateToProps = state => ({
  scalar: state.scalar,
  viewportMatrix: state.svg.filter(svg => svg.id === 'viewport').reduce((acc, val) => val.matrix, [])
})

// Viewport constants. Need to be set and leveraged both in Grid and Viewport.
const baseWidth = 1440
const baseHeight = 960
const viewboxWidth = baseWidth
const viewboxHeight = baseHeight
const halfWidth = viewboxWidth / 2
const halfHeight = viewboxHeight / 2

// @NOTE First pass code. More hardcoded into the render function than necessary.
// @TODO Set limits on scalar in the reducer to prevent excessive zoom out.
const Container = ({scalar, viewportMatrix}) => {
  // Current Center Value
  const [centerX, centerY] = viewportMatrix.slice(4, 6)
  const gridSize = 50

  // Additional grid unit added to each bound to ensure grid scales to just outside the viewport.
  // Grid unit adjusted for scalar when added, then the bould is adjusted for scalar after.
  const bounds = {
    left: (-(centerX + halfWidth) - (gridSize * scalar)) / scalar,
    right: ((halfWidth - centerX) + (gridSize * scalar)) / scalar,
    top: (-(centerY + halfHeight) - (gridSize * scalar)) / scalar,
    bottom: ((halfHeight - centerY) + (gridSize * scalar)) / scalar
  }

  // Calculate the maximum lines from the center point to any given bound direction.
  const maxLines = b => Math.ceil(b / gridSize) * gridSize

  const left = maxLines(bounds.left)
  const right = maxLines(bounds.right)
  const top = maxLines(bounds.top)
  const bottom = maxLines(bounds.bottom)

  const lines = []

  for (let n = left; n <= right; n += gridSize) {
    lines.push(<line key={`${n}-v`} x1={n} x2={n} y1={top} y2={bottom} stroke='#fff' fill='none' strokeWidth='2' />)
  }

  for (let n = top; n <= bottom; n += gridSize) {
    lines.push(<line key={`${n}-h`} x1={left} x2={right} y1={n} y2={n} stroke='#fff' fill='none' strokeWidth='2' />)
  }

  const center = (<circle cx='0' cy='0' r='6' stroke='#ddd' strokeWidth='1' fill='#ddd' />)

  // id set to viewport so that drag events fire and update the viewport matrix.
  return (
    <g id='viewport' transform={`matrix(${[1, 0, 0, 1, 0, 0].join(' ')})`}>
      {lines}
      {center}
    </g>
  )
}

export const Grid = connect(mapStateToProps)(Container)
