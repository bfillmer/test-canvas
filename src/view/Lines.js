
import React from 'react'
import {connect} from 'react-redux'

const mapStateToProps = state => ({
  lines: state.blocks.filter(block => !!block.dependsOn).map(block => {
    const a = state.svg.reduce((array, svg) => svg.id === block.dependsOn ? svg.matrix : array, [])
    const b = state.svg.reduce((array, svg) => svg.id === block.id ? svg.matrix : array, [])
    return {
      a,
      b
    }
  })
})

const point = ([x, y]) => (`${x},${y}`)

const Container = ({lines}) => {
  const renderLines = lines.map(({a, b}, key) => {
    const [aX, aY, bX, bY] = [a[4], a[5], b[4], b[5]]
    // Calculate total movement in x & y directions. [DONE]
    // Whichever is longer, move half that direction. [DONE]
    // Then the total distance in the other direction. [DONE]
    // Then finish up the rest of the way. [DONE]
    const p1 = [aX, aY]
    const [mX, mY] = [(aX + bX) / 2, (aY + bY) / 2]
    const [diffX, diffY] = [Math.abs(bX - aX), Math.abs(bY - aY)]
    const p2 = (diffX > diffY) ? [mX, aY] : [aX, mY]
    const p3 = (diffX > diffY) ? [mX, bY] : [bX, mY]
    const p4 = [bX, bY]
    const points = [
      point(p1),
      point(p2),
      point(p3),
      point(p4)
    ]
    return <polyline key={key} fill='none' stroke='#333' strokeWidth='4' strokeLinejoin='round' points={points.join(' ')} />
  })
  return (
    <g id='viewport'>
      {renderLines}
    </g>
  )
}

export const Lines = connect(mapStateToProps)(Container)
