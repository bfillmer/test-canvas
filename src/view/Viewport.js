
import React from 'react'

export const Viewport = ({
  onDragStart,
  onDragMove,
  onDragEnd,
  onWheel,
  viewportMatrix,
  children
}) => (
  <svg
    width='100%'
    height='100%'
    id='viewport'
    onMouseDown={onDragStart}
    onMouseMove={onDragMove}
    onMouseUp={onDragEnd}
    onWheel={onWheel}
    viewBox={[-(1440 / 2), -(960 / 2), 1440, 960].join(' ')}
    preserveAspectRatio='xMidYMid slice'
  >
    <g transform={`matrix(${viewportMatrix})`}>
      {children}
    </g>
  </svg>
)
