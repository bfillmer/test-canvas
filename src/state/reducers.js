// @flow
import {handleActions} from 'redux-actions'

import {types} from 'state/actions'
import {assign} from 'state/utils'
import {pan, zoom} from 'state/transformations'

import {DragStartPayload, DraggedPayload, ZoomedPayload, Svg, SvgState} from 'state/types'

export const svgObject = (
  id: string,
  type: string,
  matrix: number[] = [1, 0, 0, 1, 0, 0],
  dragX: number = 0,
  dragY: number = 0
): Svg => ({
  id,
  type,
  matrix,
  dragX,
  dragY
})

const VIEWPORT = 'VIEWPORT'
const BLOCK = 'BLOCK'

// INITIAL STATE
const initialState: SvgState = {
  dragging: false,
  scalar: 1,
  svg: [
    svgObject('viewport', VIEWPORT),
    svgObject('block-one', BLOCK, [1, 0, 0, 1, 0, 0])
  ]
}

// REDUCER
export const reducer = handleActions({
  [types.DRAG_START]: (state: SvgState, {payload}: {payload: DragStartPayload}): SvgState => {
    const {id, x, y} = payload
    return assign({}, state, {
      dragging: id,
      svg: state.svg.map((svg: Svg) => svg.id === id ? assign({}, svg, {
        dragX: x,
        dragY: y
      }) : svg)
    })
  },
  [types.DRAGGED]: (state: SvgState, {payload}: {payload: DraggedPayload}): SvgState => {
    const {x, y} = payload
    return assign({}, state, {
      svg: state.svg.map((svg: Svg) => {
        if (svg.id !== state.dragging) return svg
        const dx = (x - svg.dragX) / state.scalar
        const dy = (y - svg.dragY) / state.scalar
        return assign({}, svg, {
          matrix: pan(dx, dy, svg.matrix),
          dragX: x,
          dragY: y
        })
      })
    })
  },
  [types.DRAG_END]: (state: SvgState): SvgState => assign({}, state, {
    dragging: false
  }),
  [types.ZOOMED]: (state: SvgState, {payload}: {payload: ZoomedPayload}): SvgState => {
    const {dy, id} = payload
    const scale = dy < 0 ? 1.05 : 0.95
    const matrix = state.svg.reduce((array: number[], svg: Svg) => svg.id === id ? zoom(scale, svg.matrix) : array, [])
    return assign({}, state, {
      scalar: matrix[0],
      svg: state.svg.map((svg: Svg) => svg.id === id ? assign({}, svg, { matrix }) : svg)
    })
  }
}, initialState)
