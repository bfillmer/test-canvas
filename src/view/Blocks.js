
import React from 'react'
import {connect} from 'react-redux'

import {BLOCK} from 'state/reducers'

// @NOTE Blocks would individually connect to the store to get their normalized data, thus
// the viewport only updates when anything with positioning changes, and the internal Block
// state only updates when relevant Block data changes.
const Block = ({id, matrix}) => (
  <g transform={`matrix(${matrix.join(' ')})`}>
    <rect id={id} width='150' height='100' rx='6' ry='6' fill='#ccc' stroke='#666' strokeWidth='2' x='-75' y='-50' />
  </g>
)

const mapStateToProps = state => ({
  blocks: state.svg.filter(s => s.type === BLOCK)
})

const Container = ({blocks}) => {
  const renderBlocks = blocks.map(s => <Block key={s.id} {...s} />)
  return (<g>{renderBlocks}</g>)
}

export const Blocks = connect(mapStateToProps)(Container)
