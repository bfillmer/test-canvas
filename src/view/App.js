
import React from 'react'

import {Viewport} from 'view/Viewport'
import {Blocks} from 'view/Blocks'
import {Grid} from 'view/Grid'

export const App = () => (
  <Viewport>
    <Grid />
    <Blocks />
  </Viewport>
)
