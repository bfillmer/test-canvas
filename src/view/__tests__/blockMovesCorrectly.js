/* global test, expect */

// @TODO Generalize this whole test file so we can create new tests that use new json data, and a
// test description.
import React from 'react'
import renderer from 'react-test-renderer'
import {createStore} from 'redux'
import {Provider} from 'react-redux'

import {reducer} from 'state/reducers'

import {App} from 'view/App'

const testStore = createStore(reducer)

// Actions that got us to the expected UI visual.
import data from './moveBlockActions.json'
const actionsArray = JSON.parse(data.payload)

console.log(actionsArray)

// Replay the actions in order to end up with our needed state for our
// expected UI visual.
actionsArray.map(testStore.dispatch)

test('canvas loads and a block is moved', () => {
  const tree = renderer.create(
    <Provider store={testStore}>
      <App />
    </Provider>
  ).toJSON()
  expect(tree).toMatchSnapshot()
})
