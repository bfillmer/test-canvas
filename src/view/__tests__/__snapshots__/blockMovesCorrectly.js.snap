exports[`test canvas loads and a block is moved 1`] = `
<svg
  height="100%"
  id="viewport"
  onMouseDown={[Function]}
  onMouseMove={[Function]}
  onMouseUp={[Function]}
  onWheel={[Function]}
  preserveAspectRatio="xMidYMid slice"
  viewBox="-720 -480 1440 960"
  width="100%">
  <g
    transform="matrix(1 0 0 1 0 0)">
    <g
      id="viewport"
      transform="matrix(1 0 0 1 0 0)">
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-750}
        x2={-750}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-700}
        x2={-700}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-650}
        x2={-650}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-600}
        x2={-600}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-550}
        x2={-550}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-500}
        x2={-500}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-450}
        x2={-450}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-400}
        x2={-400}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-350}
        x2={-350}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-300}
        x2={-300}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-250}
        x2={-250}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-200}
        x2={-200}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-150}
        x2={-150}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-100}
        x2={-100}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-50}
        x2={-50}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={0}
        x2={0}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={50}
        x2={50}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={100}
        x2={100}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={150}
        x2={150}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={200}
        x2={200}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={250}
        x2={250}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={300}
        x2={300}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={350}
        x2={350}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={400}
        x2={400}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={450}
        x2={450}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={500}
        x2={500}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={550}
        x2={550}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={600}
        x2={600}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={650}
        x2={650}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={700}
        x2={700}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={750}
        x2={750}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={800}
        x2={800}
        y1={-500}
        y2={550} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-750}
        x2={800}
        y1={-500}
        y2={-500} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-750}
        x2={800}
        y1={-450}
        y2={-450} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-750}
        x2={800}
        y1={-400}
        y2={-400} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-750}
        x2={800}
        y1={-350}
        y2={-350} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-750}
        x2={800}
        y1={-300}
        y2={-300} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-750}
        x2={800}
        y1={-250}
        y2={-250} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-750}
        x2={800}
        y1={-200}
        y2={-200} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-750}
        x2={800}
        y1={-150}
        y2={-150} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-750}
        x2={800}
        y1={-100}
        y2={-100} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-750}
        x2={800}
        y1={-50}
        y2={-50} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-750}
        x2={800}
        y1={0}
        y2={0} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-750}
        x2={800}
        y1={50}
        y2={50} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-750}
        x2={800}
        y1={100}
        y2={100} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-750}
        x2={800}
        y1={150}
        y2={150} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-750}
        x2={800}
        y1={200}
        y2={200} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-750}
        x2={800}
        y1={250}
        y2={250} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-750}
        x2={800}
        y1={300}
        y2={300} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-750}
        x2={800}
        y1={350}
        y2={350} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-750}
        x2={800}
        y1={400}
        y2={400} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-750}
        x2={800}
        y1={450}
        y2={450} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-750}
        x2={800}
        y1={500}
        y2={500} />
      <line
        fill="none"
        stroke="#fff"
        strokeWidth="2"
        x1={-750}
        x2={800}
        y1={550}
        y2={550} />
      <circle
        cx="0"
        cy="0"
        fill="#ddd"
        r="6"
        stroke="#ddd"
        strokeWidth="1" />
    </g>
    <g
      id="viewport">
      <polyline
        fill="none"
        points="50,200 50,94 -121,94 -121,-12"
        stroke="#333"
        strokeLinejoin="round"
        strokeWidth="4" />
      <polyline
        fill="none"
        points="-121,-12 -10.5,-12 -10.5,-200 100,-200"
        stroke="#333"
        strokeLinejoin="round"
        strokeWidth="4" />
    </g>
    <g>
      <g
        transform="matrix(1 0 0 1 -121 -12)">
        <rect
          fill="#ccc"
          height="100"
          id="block-one"
          rx="6"
          ry="6"
          stroke="#666"
          strokeWidth="2"
          width="150"
          x="-75"
          y="-50" />
      </g>
      <g
        transform="matrix(1 0 0 1 50 200)">
        <rect
          fill="#ccc"
          height="100"
          id="block-two"
          rx="6"
          ry="6"
          stroke="#666"
          strokeWidth="2"
          width="150"
          x="-75"
          y="-50" />
      </g>
      <g
        transform="matrix(1 0 0 1 100 -200)">
        <rect
          fill="#ccc"
          height="100"
          id="block-three"
          rx="6"
          ry="6"
          stroke="#666"
          strokeWidth="2"
          width="150"
          x="-75"
          y="-50" />
      </g>
    </g>
  </g>
</svg>
`;
